### PR TITLE
Typing Stampy

### PR DESCRIPTION
**Pull request by tayler6000**
Lovkush and I are typing Stampy.

Todo:

- [ ] api
- [ ] database
- [ ] images
- [ ] modules
- [ ] scripts
- [ ] servicemodules
- [ ] test
- [ ] utilities

### CODE REVIEW

**Comment by Lovkush-A** [comment]
@tayler6000 @MatthewBaggins @Aprillion 

requesting a review to get this little bits of changes in codebase. rather than having one huge pr for all typehints.

File: utilities/discordutils.py [hunk 1]
```diff
@@ -28,29 +28,34 @@ def parse_discord_roles(self, roles: list[discord.Role]) -> None:
 
 class DiscordChannel(ServiceChannel):
     def __init__(self, channel: discord.abc.Messageable, server: Optional[ServiceServer]):
-        self._channel = channel
+        self._channel: discord.abc.Messageable = channel
         self.history = channel.history
-        if not isinstance(channel, discord.DMChannel):
+        if not isinstance(channel, discord.abc.Messageable):
```
Comment (Aprillion) [inline utilities/discordutils.py:33]: @tayler6000  what is this about, please?  either `channel` parameter has some type and assignment on line 31 should infer the type and this condition on line 33 doesn't make any sense at all

..or the parameter type is incorrect and `channel.id` and `channel.name` are both unsafe access - `not isinstance(...)` will NOT make it safe - so why this particular condition?

File: utilities/discordutils.py [hunk 2]
```diff
@@ -60,12 +65,17 @@ def __init__(self, msg: discord.message.Message):
 
         super().__init__(str(msg.id), msg.content, author, channel, service)
         self.clean_content = msg.clean_content.replace("\u200b", "")
+        self.mentions: list[DiscordUser]
         self._parse_discord_mentions(msg.mentions)
-        self.reference = msg.reference
+        self.reference = None
+        if msg.reference:
+            if msg.reference.resolved:
```
Comment (Aprillion) [inline utilities/discordutils.py:72]:
```suggestion
        if msg.reference and msg.reference.resolved:
```

File: utilities/serviceutils.py [hunk 3]
```diff
@@ -103,8 +103,7 @@ def __hash__(self):
         """
         return hash(self.id) >> 22
 
-    async def send(self, content):
-        raise NotImplementedError()
+    async def send(self, *args, **kwargs) -> Any: ...
```
Comment (Aprillion) [inline utilities/serviceutils.py:106]: why not `-> None` or whatever should be the actual type?

Comment (MatthewBaggins) [reply utilities/serviceutils.py:106]: `NoReturn` I think
https://docs.python.org/3/library/typing.html#typing.NoReturn

File: utilities/discordutils.py (hunk 1 above)
Comment (Lovkush-A) [reply utilities/discordutils.py:33]: Looks like tayler is awol. What do you think resolution of this should be? Should we remove the isinstance checks for `channel` and just do:

```python
channel_id = channel.id
name = ""
```

File: utilities/discordutils.py (hunk 2 above)
Comment (Lovkush-A) [reply utilities/discordutils.py]: done

File: utilities/serviceutils.py (hunk 3 above)
Comment (Lovkush-A) [reply utilities/serviceutils.py:106]: we could create a alias which is union of the different types we expect for this?

**Review by Aprillion — APPROVED** [review]

File: utilities/serviceutils.py (hunk 3 above)
Comment (Lovkush-A) [reply utilities/serviceutils.py]: created alias

**Pull request merged**